modules/junit-jupiter/src/main/java/org/testcontainers/junit/jupiter/TestcontainersExtension.java

```diff
@@ -0,0 +1,125 @@
+import org.testcontainers.lifecycle.Startable;
+
+import java.lang.reflect.Field;
+import java.util.LinkedHashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+class TestcontainersExtension implements TestInstancePostProcessor, BeforeEachCallback {
+
+    private static final Namespace NAMESPACE = Namespace.create(TestcontainersExtension.class);
+
+    private static final String TEST_INSTANCE = "testInstance";
+
+    @Override
+    public void postProcessTestInstance(final Object testInstance, final ExtensionContext context) {
+        ExtensionContext.Store store = context.getStore(NAMESPACE);
+        store.put(TEST_INSTANCE, testInstance);
+
+        findSharedContainers(testInstance)
+            .forEach(adapter -> store.getOrComputeIfAbsent(adapter.key, k -> adapter.start()));
+    }
+
+    @Override
+    public void beforeEach(final ExtensionContext context) {
+        collectParentTestInstances(context)
+            .parallelStream()
+            .flatMap(this::findRestartContainers)
+            .forEach(adapter -> context.getStore(NAMESPACE)
+                .getOrComputeIfAbsent(adapter.key, k -> adapter.start()));
+    }
+
+    private Set<Object> collectParentTestInstances(final ExtensionContext context) {
+        Set<Object> testInstances = new LinkedHashSet<>();
+        Optional<ExtensionContext> current = Optional.of(context);
+        while(current.isPresent()) {
+            ExtensionContext ctx = current.get();
+            Object testInstance = ctx.getStore(NAMESPACE).remove(TEST_INSTANCE);
+            if (testInstance != null) {
+                testInstances.add(testInstance);
+            }
+            current = ctx.getParent();
+        }
+        return testInstances;
+    }
+
+    private Stream<StoreAdapter> findSharedContainers(Object testInstance) {
+        return ReflectionUtils.findFields(
+                testInstance.getClass(),
+                isSharedContainer(),
+                ReflectionUtils.HierarchyTraversalMode.TOP_DOWN)
+            .stream()
+            .map(f -> getContainerInstance(testInstance, f));
+    }
+
+    private Predicate<Field> isSharedContainer() {
+        return isContainer().and(ReflectionUtils::isStatic);
+    }
+
+    private Stream<StoreAdapter> findRestartContainers(Object testInstance) {
+        return ReflectionUtils.findFields(
+                testInstance.getClass(),
+                isRestartContainer(),
+                ReflectionUtils.HierarchyTraversalMode.TOP_DOWN)
+            .stream()
+            .map(f -> getContainerInstance(testInstance, f));
+    }
+
+    private Predicate<Field> isRestartContainer() {
+        return isContainer().and(ReflectionUtils::isNotStatic);
+    }
+
+    private static Predicate<Field> isContainer() {
+        return field -> Startable.class.isAssignableFrom(field.getType()) && AnnotationSupport.isAnnotated(field, Container.class);
```

Comment:
I would return any field annotated with `@Container` and throw and error if it is not `Startable`, otherwise it will simply ignore it